base/compiler/abstractinterpretation.jl

```diff
@@ -1174,7 +1182,8 @@ function abstract_invoke(interp::AbstractInterpreter, argtypes::Vector{Any}, sv:
     nargtype === Bottom && return CallMeta(Bottom, false)
     nargtype isa DataType || return CallMeta(Any, false) # other cases are not implemented below
     isdispatchelem(ft) || return CallMeta(Any, false) # check that we might not have a subtype of `ft` at runtime, before doing supertype lookup below
-    types = rewrap_unionall(Tuple{ft, unwrap_unionall(types).parameters...}, types)
+    ft = ft::DataType
```

Comment:
it could also be `typeof(Bottom)` here

Reply:
Yeah, now I'm wondering why we can have this union type `Union{typeof(Bottom),DataType}` and at the same time `typof(Bottom)::DataType` succeeds ? If `typeof(Bottom)` can be used in place of `DataType`, I think it might be better to not form `Union{typeof(Bottom),DataType}`.

Reply:
`typeof(Bottom) isa DataType`, but `!(typeof(Bottom) <: DataType)`

Reply:
Thanks @martinholters .
Then I think this type assertion is okay, that should be better than `ft :: Any`

Reply:
okay, true